File: AdventOfCode2022/Python/Day08/Day08.py
import json

def get_inputs(file):
    """
    function to read in list of ints from a txt file
    """

    with open(file) as f:
        data = f.read().split('\n')

    return data

def Part1(filename):
    """
    Solutions to Part 1
    """
    data = get_inputs(filename)

    # count trees
    interior_trees = 0
    for i in range(0,len(data)):
        row_of_trees = data[i]
        for j in range(0, len(row_of_trees)):
            tree_being_checked = int(data[i][j])

            # visibility from top
            top_check = True
            for y in range(0,i):
                if int(data[y][j]) >= tree_being_checked:
                    top_check = False
                    
            # visibility from bottom
            bot_check = True
            for y in range(i+1, len(data)):
                if int(data[y][j]) >= tree_being_checked:
                    bot_check = False

            # visibility from left
            lef_check = True
            for x in range(0, j):
                if int(data[i][x]) >= tree_being_checked:
                    lef_check = False

            # visibility from right
            rig_check = True
            for x in range(j+1, len(data[0])):
                if int(data[i][x]) >= tree_being_checked:
                    rig_check = False
            
            if any ([top_check, bot_check, lef_check, rig_check]):
                interior_trees += 1

    ans = interior_trees #- edge_trees
    return ans
    

def Part2(filename):
    """
    Solutions to Part 2
    """
    data = get_inputs(filename)

    # count trees
    scenic_scores = []
    for i in range(1,len(data)-1):
        row_of_trees = data[i]
        for j in range(1, len(row_of_trees)-1):
            tree_being_checked = int(data[i][j])

            # visibility from top
            top_check = True
            top_view = 0
            top_list = []
            # not sure why python wouldnt let me use a descending range
            for y in range(0,i):
                top_list.append(data[y][j])
            top_list.reverse()
            for d in top_list:
                top_view += 1
                if int(d) >= tree_being_checked:                    
                    top_check = False
                    break
                    
            # visibility from bottom
            bot_check = True
            bot_view = 0
            for y in range(i+1, len(data)):
                bot_view += 1
                if int(data[y][j]) >= tree_being_checked:                    
                    bot_check = False
                    break

            # visibility from left
            lef_check = True
            lef_view = 0
            lef_list = []
            for x in range(0,j):
                lef_list.append(data[i][x])
            lef_list.reverse()
            for d in lef_list:
                lef_view += 1
                if int(d) >= tree_being_checked:
                    lef_check = False
                    break

            # visibility from right
            rig_check = True
            rig_view = 0
            for x in range(j+1, len(data[0])):
                rig_view += 1
                
                if int(data[i][x]) >= tree_being_checked:
                    rig_check = False
                    break
            
            if any ([top_check, bot_check, lef_check, rig_check]):
                scenic_scores.append(top_view * bot_view * lef_view * rig_view)

    ans = max(scenic_scores)
    return ans

def main():
    """
    main function
    """ 
    testfile = "tests.txt"
    inputfile = "inputs.txt"

    print("Advent-Of-Code 2022 - Day08")

    print(f'\n"{testfile}"')
    print(f' Answer to Part 1 = {Part1(testfile)}')
    print(f' Answer to Part 2 = {Part2(testfile)}')

    print(f'\n"{inputfile}"')
    print(f' Answer to Part 1 = {Part1(inputfile)}')
    print(f' Answer to Part 2 = {Part2(inputfile)}')

if __name__ == "__main__":
    main()

File: AdventOfCode2022/Go/Day08/Day08.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

func Parse(filename string) [][]int {

	file, _ := os.Open(filename)
	fscanner := bufio.NewScanner(file)

	data := [][]int{}

	for fscanner.Scan() {
		line := fscanner.Text()

		lineData := []int{}
		for i := 0; i < len(line); i++ {
			char, _ := strconv.Atoi(string(line[i]))
			lineData = append(lineData, char)
		}
		data = append(data, lineData)
	}

	return data
}

func Part1(filename string) int {

	data := Parse(filename)

	numOfVisible := (2 * len(data)) + (2 * len(data[0])) - 4

	for y := 1; y < len(data)-1; y++ {
		for x := 1; x < len(data[y])-1; x++ {

			visibleUp := true
			visibleDown := true
			visibleLeft := true
			visibleRight := true

			for lc := 0; lc < x; lc++ {
				if data[y][lc] >= data[y][x] {
					visibleLeft = false
				}
			}

			for rc := x + 1; rc < len(data[x]); rc++ {
				if data[y][rc] >= data[y][x] {
					visibleRight = false
				}
			}

			for uc := 0; uc < y; uc++ {
				if data[uc][x] >= data[y][x] {
					visibleUp = false
				}
			}

			for dc := y + 1; dc < len(data); dc++ {
				if data[dc][x] >= data[y][x] {
					visibleDown = false
				}
			}

			if visibleLeft || visibleRight || visibleUp || visibleDown {
				numOfVisible += 1
			}
		}
	}
	return numOfVisible
}

func Part2(filename string) int {

	data := Parse(filename)

	scenicScore := 0

	for y := 1; y < len(data)-1; y++ {
		for x := 1; x < len(data[y])-1; x++ {

			distanceLeft := 0
			distanceRight := 0
			distanceUp := 0
			distanceDown := 0

			for lc := x - 1; lc >= 0; lc-- {
				distanceLeft++
				if data[y][lc] >= data[y][x] {
					break
				}
			}

			for rc := x + 1; rc < len(data[x]); rc++ {
				distanceRight++
				if data[y][rc] >= data[y][x] {
					break
				}
			}

			for uc := y - 1; uc >= 0; uc-- {
				distanceUp++
				if data[uc][x] >= data[y][x] {
					break
				}
			}

			for dc := y + 1; dc < len(data); dc++ {
				distanceDown++
				if data[dc][x] >= data[y][x] {
					break
				}
			}

			currentScenicScore := distanceLeft * distanceRight * distanceUp * distanceDown
			if currentScenicScore > scenicScore {
				scenicScore = currentScenicScore
			}
		}
	}
	return scenicScore
}

func main() {

	testfile := "tests.txt"
	inputfile := "inputs.txt"

	fmt.Println("Advent-Of-Code 2022 - Day06")
	fmt.Printf("Tests : Answer to Part 1 = %v\n", Part1(testfile))
	fmt.Printf("Inputs: Answer to Part 1 = %v\n", Part1(inputfile))
	fmt.Printf("Tests : Answer to Part 2 = %v\n", Part2(testfile))
	fmt.Printf("Inputs: Answer to Part 2 = %v\n", Part2(inputfile))
}
